punchcut/src/source/glyf/deltas.rs

```diff
@@ -0,0 +1,289 @@
+    glyph: SimpleGlyph,
+    working_points: &mut [Point<Fixed>],
+    deltas: &mut [Point<Fixed>],
+) -> Result<(), ReadError> {
+    for delta in deltas.iter_mut() {
+        *delta = Default::default();
+    }
+    let Ok(var_data) = gvar.glyph_variation_data(glyph_id) else {
+        // Empty variation data for a glyph is not an error.
+        return Ok(());
+    };
+    let SimpleGlyph {
+        points,
+        flags,
+        contours,
+    } = glyph;
+    // Include the first phantom point if the font is missing variable metrics
+    // for left side bearings. The adjustment made here may affect the final
+    // shift of the outline.
+    let actual_len = if has_var_lsb {
+        points.len() - 4
+    } else {
+        points.len() - 3
+    };
+    for tuple in var_data.tuples() {
+        let Some(scalar) = tuple.compute_scalar(coords) else {
+            continue;
+        };
+        if tuple.all_points() {
+            // When a tuple contains all points, we can simply accumulate the deltas directly.
+            for (delta, tuple_delta) in deltas[0..actual_len].iter_mut().zip(tuple.deltas()) {
+                *delta += tuple_delta.apply_scalar(scalar);
+            }
+        } else {
+            // Otherwise, we need to infer missing deltas by interpolation.
+            // Prepare our working buffer by converting the points to 16.16
+            // and clearing the HAS_DELTA flags.
+            for ((flag, point), working_point) in
+                flags.iter_mut().zip(points).zip(&mut working_points[..])
+            {
+                *working_point = point.map(Fixed::from_i32);
+                flag.clear_marker(PointMarker::HAS_DELTA);
+            }
+            for tuple_delta in tuple.deltas() {
+                let ix = tuple_delta.position as usize;
+                if let (Some(flag), Some(point)) = (flags.get_mut(ix), working_points.get_mut(ix)) {
+                    flag.set_marker(PointMarker::HAS_DELTA);
+                    *point += tuple_delta.apply_scalar(scalar);
+                }
+            }
+            interpolate_deltas(points, flags, contours, &mut working_points[..])
+                .ok_or(ReadError::OutOfBounds)?;
+            for ((delta, point), working_point) in deltas[..actual_len]
+                .iter_mut()
+                .zip(points)
+                .zip(working_points.iter())
+            {
+                *delta += *working_point - point.map(Fixed::from_i32);
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Interpolate points without delta values, similar to the IUP hinting
+/// instruction.
+///
+/// Modeled after the FreeType implementation: <https://github.com/freetype/freetype/blob/bbfcd79eacb4985d4b68783565f4b494aa64516b/src/truetype/ttgxvar.c#L3881>
+fn interpolate_deltas(
+    points: &[Point<i32>],
+    flags: &[PointFlags],
+    contours: &[u16],
+    out_points: &mut [Point<Fixed>],
+) -> Option<()> {
+    if contours.is_empty() {
+        return Some(());
+    }
```

Comment:
since the entirety of your code other than setting point_ix to 0 is in the loop over contours perhaps the guard clause is unnecessary?

Reply:
Good catch. Removed the worthless early exit.